### PR TITLE
FMWK-196 Index Cardinality

### PR DESCRIPTION
**Pull request by roimenashe**

### CODE REVIEW

File: src/main/java/org/springframework/data/aerospike/query/StatementBuilder.java [hunk 1]
```diff
@@ -59,13 +64,44 @@ public Statement build(String namespace, String set, @Nullable Query query, Stri
...
+                if (currBinValuesRatio < minBinValuesRatio && currBinValuesRatio != 0) {
+                    minBinValuesRatio = currBinValuesRatio;
+                    minBinValuesRatioQualifier = innerQualifier;
+                }
+            }
+        }
+
+        // If index with min bin values ratio found, set filter with the matching qualifier
+        if (minBinValuesRatioQualifier != null) {
+            Filter filter = minBinValuesRatioQualifier.setQueryAsFilter();
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/StatementBuilder.java:99]: The `setFilterFromSingleQualifier` method can be reused here.

File: src/main/java/org/springframework/data/aerospike/query/StatementBuilder.java [hunk 2]
```diff
@@ -101,4 +139,20 @@ private boolean isIndexedBin(Statement stmt, Qualifier qualifier) {
...
+
+        // Return the lowest bin values ratio of the indexes in indexList
+        Optional<Index> minBinValuesRatio = indexList.stream()
+            .filter(index -> index.getBinValuesRatio() != 0)
+            .min(Comparator.comparing(Index::getBinValuesRatio));
+
+        if (minBinValuesRatio.isPresent()) {
+            return minBinValuesRatio.get().getBinValuesRatio();
+        }
+        return 0;
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/StatementBuilder.java:156]: refactor to:
```java
return minBinValuesRatio.map(Index::getBinValuesRatio).orElse(0);
```

File: src/main/java/org/springframework/data/aerospike/query/model/Index.java [hunk 3]
```diff
@@ -19,30 +19,35 @@
...
-    CTX[] ctx;
+    private final String name;
+    private final String namespace;
+    private final String set;
+    private final String bin;
+    private final IndexType indexType;
+    private final IndexCollectionType indexCollectionType;
+    private final CTX[] ctx;
+    @Setter
+    private Integer binValuesRatio;
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/model/Index.java:50]: Can be `int`?

File: src/main/java/org/springframework/data/aerospike/query/cache/InternalIndexOperations.java [hunk 4]
```diff
@@ -37,6 +44,8 @@ public class InternalIndexOperations {
 
     private final IndexInfoParser indexInfoParser;
 
+    private final Logger log = LoggerFactory.getLogger(InternalIndexOperations.class);
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/cache/InternalIndexOperations.java:47]: Use the `@Slf4j` annotation instead.

File: src/main/java/org/springframework/data/aerospike/query/StatementBuilder.java [hunk 5]
```diff
@@ -59,13 +64,44 @@ public Statement build(String namespace, String set, @Nullable Query query, Stri
...
+    private void setFilterFromMultipleQualifiers(Statement stmt, Qualifier qualifier) {
+        int minBinValuesRatio = Integer.MAX_VALUE;
+        Qualifier minBinValuesRatioQualifier = null;
+
+        for (Qualifier innerQualifier : qualifier.getQualifiers()) {
+            if (innerQualifier != null && isIndexedBin(stmt, innerQualifier)) {
+                int currBinValuesRatio = getMinBinValuesRatioForQualifier(stmt, innerQualifier);
+                // Compare the cardinality of each qualifier and select the qualifier that has the index with
+                // the lowest bin values ratio
+                if (currBinValuesRatio < minBinValuesRatio && currBinValuesRatio != 0) {
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/StatementBuilder.java:90]: I would move `currBinValuesRatio != 0` to be first in this statement.

File: src/main/java/org/springframework/data/aerospike/query/cache/InternalIndexOperations.java [hunk 6]
```diff
@@ -60,4 +69,35 @@ public IndexesInfo parseIndexesInfo(String infoResponse) {
...
+                    Arrays.stream(indexStatData.split(";"))
+                        .map(String::trim)
+                        .toList().stream()
+                        .map(stat -> Arrays.stream(stat.split("="))
+                            .map(String::trim)
+                            .collect(Collectors.toList()))
+                        .collect(Collectors.toMap(t -> t.get(0), t -> t.get(1)))
+                        .get("entries_per_bval"));
+            } catch (Exception e) {
+                log.warn("Failed to fetch secondary index %s cardinality".formatted(indexName), e);
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/query/cache/InternalIndexOperations.java:98]: Use
```java
log.warn("Failed to fetch secondary index {} cardinality", indexName, e);
```
instead.

**Review by reugn — APPROVED** [review]

**Review by agrgr — APPROVED** [review]

**Pull request merged**